src/dotnet/commands/dotnet-build/DotnetProjectBuilder.cs

```diff
@@ -150,118 +146,39 @@ protected override CompilationResult RunCompile(ProjectGraphNode projectNode)
-            {
-                // this is the first compilation; cannot tell if CLI changed
-                return false;
-            }
+            PrintIncrementalResult(graphNode.ProjectContext.GetDisplayName(), result);
 
-            var currentContent = ComputeCurrentVersionFileData();
-
-            var versionsAreEqual = string.Equals(currentContent, File.ReadAllText(versionFileFromLastCompile), StringComparison.OrdinalIgnoreCase);
-
-            return !versionsAreEqual;
+            return result.NeedsRebuilding;
         }
 
-        protected override bool NeedsRebuilding(ProjectGraphNode graphNode)
+        private void PrintIncrementalResult(string projectName, IncrementalResult result)
         {
-            var project = graphNode.ProjectContext;
-            if (_args.ShouldNotUseIncrementality)
+            if (result.NeedsRebuilding)
             {
-                return true;
-            }
-            if (!_args.ShouldSkipDependencies &&
-                graphNode.Dependencies.Any(d => GetCompilationResult(d) != CompilationResult.IncrementalSkip))
-            {
-                Reporter.Output.WriteLine($"Project {project.GetDisplayName()} will be compiled because some of it's dependencies changed");
-                return true;
-            }
-            var preconditions = _preconditionManager.GetIncrementalPreconditions(graphNode);
-            if (preconditions.PreconditionsDetected())
-            {
-                return true;
-            }
-
-            if (CLIChangedSinceLastCompilation(project))
-            {
-                Reporter.Output.WriteLine($"Project {project.GetDisplayName()} will be compiled because the version or bitness of the CLI changed since the last build");
-                return true;
-            }
-
-            var compilerIO = _compilerIOManager.GetCompileIO(graphNode);
-
-            // rebuild if empty inputs / outputs
-            if (!(compilerIO.Outputs.Any() && compilerIO.Inputs.Any()))
-            {
-                Reporter.Output.WriteLine($"Project {project.GetDisplayName()} will be compiled because it either has empty inputs or outputs");
-                return true;
-            }
-
-            //rebuild if missing inputs / outputs
-            if (_compilerIOManager.AnyMissingIO(project, compilerIO.Outputs, "outputs") || _compilerIOManager.AnyMissingIO(project, compilerIO.Inputs, "inputs"))
-            {
-                return true;
-            }
-
-            // find the output with the earliest write time
-            var minOutputPath = compilerIO.Outputs.First();
-            var minDateUtc = File.GetLastWriteTimeUtc(minOutputPath);
-
-            foreach (var outputPath in compilerIO.Outputs)
-            {
-                if (File.GetLastWriteTimeUtc(outputPath) >= minDateUtc)
+                Reporter.Output.WriteLine($"Project {projectName} will be compiled because {result.Reason}");
+                foreach (var item in result.Items)
```

Comment:
Maybe it makes sense to check for Verbose before even getting into this loop?  That way if Verbose isn't turned on, you don't need to loop through the Items at all.